src/preamble.js

```diff
@@ -2221,6 +2221,19 @@ function integrateWasmJS(Module) {
       Module['asm'] = exports;
       Module["usingWasm"] = true;
     }
+
+    // User shell pages can write their own Module.instantiateWasm = function(imports, successCallback) callback
+    // to manually instantiate the Wasm module themselves. This allows pages to run the instantiation parallel
+    // to any other async startup actions they are performing.
+    if (Module['instantiateWasm']) {
+      try {
+        return Module['instantiateWasm'](info, receiveInstance);
```

Comment:
Looks good, thanks.

One thing that occurred to me while reading that, is that even in the synchronous mode it will add and remove a dep. Worth checking that doesn't hit a bug (where we get to 0 deps and so try to call main, perhaps too early).